src/providers/db2/qgsdb2expressioncompiler.cpp

```diff
@@ -0,0 +1,260 @@
+  Adapted from MSSQL provider by Tamas Szekeres
+****************************************************************************
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ ***************************************************************************/
+
+#include "qgsdb2expressioncompiler.h"
+
+QgsDb2ExpressionCompiler::QgsDb2ExpressionCompiler( QgsDb2FeatureSource* source )
+    : QgsSqlExpressionCompiler( source->mFields
+                              )
+{
+
+}
+
+QString nodeType( const QgsExpression::Node* node )
+{
+  QString opString = "?";
+  if ( node->nodeType() == QgsExpression::ntUnaryOperator ) opString =  "ntUnaryOperator";
+  if ( node->nodeType() == QgsExpression::ntBinaryOperator ) opString =  "ntBinaryOperator";
+  if ( node->nodeType() == QgsExpression::ntInOperator ) opString =  "ntInOperator";
+  if ( node->nodeType() == QgsExpression::ntFunction ) opString =  "ntFunction";
+  if ( node->nodeType() == QgsExpression::ntLiteral ) opString =  "ntLiteral";
+  if ( node->nodeType() == QgsExpression::ntColumnRef ) opString =  "ntColumnRef";
+  if ( node->nodeType() == QgsExpression::ntCondition ) opString =  "ntCondition";
+  QString result = QString( "%1 - " ).arg( node->nodeType() ) + opString;
+  return result;
+
+}
+
+QString resultType( QgsSqlExpressionCompiler::Result result )
+{
+  if ( result == QgsSqlExpressionCompiler::None ) return "None";
+  if ( result == QgsSqlExpressionCompiler::Complete ) return "Complete";
+  if ( result == QgsSqlExpressionCompiler::Partial ) return "Partial";
+  if ( result == QgsSqlExpressionCompiler::Fail ) return "Fail";
+  return "Other result";
+
+}
+
+QgsSqlExpressionCompiler::Result QgsDb2ExpressionCompiler::compileNode( const QgsExpression::Node* node, QString& result )
+{
+  QgsDebugMsg( QString( "nodeType: %1" ).arg( nodeType( node ) ) );
+  if ( node->nodeType() == QgsExpression::ntColumnRef )
+  {
+    const QgsExpression::NodeColumnRef *n( static_cast<const QgsExpression::NodeColumnRef*>( node ) );
+    QgsDebugMsg( QString( "column ref node: " ) + n->dump() );
+    // TODO - consider escaped names - not sure how to handle
+    QString upperName = n->name().toUpper();
+    int idx = mFields.indexFromName( upperName );
+    QgsDebugMsg( QString( "%1 - %2" ).arg( idx ).arg( upperName ) );
+    if ( idx > -1 )
+    {
+      result = upperName;
+      QgsDebugMsg( "return Complete" );
+      return Complete;
+    }
+    QgsDebugMsg( "return Fail" );
+    return Fail;
+  }
+  if ( node->nodeType() == QgsExpression::ntLiteral )
+  {
+    const QgsExpression::NodeLiteral* n = static_cast<const QgsExpression::NodeLiteral*>( node );
+
+    bool ok = false;
+    if ( n->dump().toUpper() == "NULL" ) // expression compiler doesn't handle this correctly
```

Comment:
I believe the problem is that an expression like
column is null
was being returned as
column is 'null'
which doesn't work for DB2.